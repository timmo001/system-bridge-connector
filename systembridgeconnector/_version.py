"""
Provides systembridgeconnector version information.
"""

# This file is auto-generated! Do not edit!
# Use `python -m incremental.update systembridgeconnector` to change this file.

from incremental import Version

__version__ = Version("systembridgeconnector", 4, 0, 6, dev=0)
__all__ = ["__version__"]
