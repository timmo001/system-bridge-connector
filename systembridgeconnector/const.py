"""Constants."""

from enum import StrEnum


class QueryParameter(StrEnum):
    """Query Parameter."""

    ALBUM = "album"
    API_PORT = "apiPort"
    ARTIST = "artist"
    AUTOPLAY = "autoplay"
    BASE = "base"
    FILENAME = "filename"
    PATH = "path"
    TITLE = "title"
    TOKEN = "token"
    URL = "url"
    VOLUME = "volume"


class EventKey(StrEnum):
    """Event Key."""

    APP_ICON = "app_icon"
    APP_NAME = "app_name"
    BASE = "base"
    DATA = "data"
    DIRECTORIES = "directories"
    EVENT = "event"
    FILE = "file"
    FILENAME = "filename"
    FILES = "files"
    ID = "id"
    KEY = "key"
    MESSAGE = "message"
    MODULE = "module"
    MODULES = "modules"
    PATH = "path"
    SETTING = "setting"
    SUBTYPE = "subtype"
    TEXT = "text"
    TIMEOUT = "timeout"
    TITLE = "title"
    TOKEN = "token"
    TYPE = "type"
    URL = "url"
    VALUE = "value"
    VERSION = "version"
    VERSIONS = "versions"


class EventType(StrEnum):
    """Event Type."""

    APPLICATION_UPDATE = "APPLICATION_UPDATE"
    APPLICATION_UPDATING = "APPLICATION_UPDATING"
    DATA_GET = "DATA_GET"
    DATA_LISTENER_REGISTERED = "DATA_LISTENER_REGISTERED"
    DATA_LISTENER_UNREGISTERED = "DATA_LISTENER_UNREGISTERED"
    DATA_UPDATE = "DATA_UPDATE"
    DIRECTORIES = "DIRECTORIES"
    ERROR = "ERROR"
    EXIT_APPLICATION = "EXIT_APPLICATION"
    FILE = "FILE"
    FILES = "FILES"
    GET_DATA = "GET_DATA"
    GET_DIRECTORIES = "GET_DIRECTORIES"
    GET_FILE = "GET_FILE"
    GET_FILES = "GET_FILES"
    GET_SETTINGS = "GET_SETTINGS"
    KEYBOARD_KEY_PRESSED = "KEYBOARD_KEY_PRESSED"
    KEYBOARD_KEYPRESS = "KEYBOARD_KEYPRESS"
    KEYBOARD_TEXT = "KEYBOARD_TEXT"
    KEYBOARD_TEXT_SENT = "KEYBOARD_TEXT_SENT"
    MEDIA_CONTROL = "MEDIA_CONTROL"
    NOTIFICATION = "NOTIFICATION"
    NOTIFICATION_SENT = "NOTIFICATION_SENT"
    OPEN = "OPEN"
    OPENED = "OPENED"
    POWER_HIBERNATE = "POWER_HIBERNATE"
    POWER_HIBERNATING = "POWER_HIBERNATING"
    POWER_LOCK = "POWER_LOCK"
    POWER_LOCKING = "POWER_LOCKING"
    POWER_LOGGINGOUT = "POWER_LOGGINGOUT"
    POWER_LOGOUT = "POWER_LOGOUT"
    POWER_RESTART = "POWER_RESTART"
    POWER_RESTARTING = "POWER_RESTARTING"
    POWER_SHUTDOWN = "POWER_SHUTDOWN"
    POWER_SHUTTINGDOWN = "POWER_SHUTTINGDOWN"
    POWER_SLEEP = "POWER_SLEEP"
    POWER_SLEEPING = "POWER_SLEEPING"
    REGISTER_DATA_LISTENER = "REGISTER_DATA_LISTENER"
    SETTINGS_UPDATED = "SETTINGS_UPDATED"
    SETTINGS_RESULT = "SETTINGS_RESULT"
    UNREGISTER_DATA_LISTENER = "UNREGISTER_DATA_LISTENER"
    UPDATE_SETTINGS = "UPDATE_SETTINGS"


class EventSubType(StrEnum):
    """Event SubType."""

    BAD_TOKEN = "BAD_TOKEN"
    BAD_DIRECTORY = "BAD_DIRECTORY"
    BAD_FILE = "BAD_FILE"
    BAD_JSON = "BAD_JSON"
    BAD_KEY = "MISSING_KEY"
    BAD_PATH = "BAD_PATH"
    BAD_REQUEST = "BAD_REQUEST"
    INVALID_ACTION = "INVALID_ACTION"
    LISTENER_ALREADY_REGISTERED = "LISTENER_ALREADY_REGISTERED"
    LISTENER_NOT_REGISTERED = "LISTENER_NOT_REGISTERED"
    MISSING_ACTION = "MISSING_ACTION"
    MISSING_TOKEN = "MISSING_TOKEN"
    MISSING_BASE = "MISSING_BASE"
    MISSING_KEY = "MISSING_KEY"
    MISSING_MODULES = "MISSING_MODULES"
    MISSING_PATH = "MISSING_PATH"
    MISSING_PATH_URL = "MISSING_PATH_URL"
    MISSING_SETTING = "MISSING_SETTING"
    MISSING_TEXT = "MISSING_TEXT"
    MISSING_TITLE = "MISSING_TITLE"
    MISSING_VALUE = "MISSING_VALUE"
    UNKNOWN_EVENT = "UNKNOWN_EVENT"
