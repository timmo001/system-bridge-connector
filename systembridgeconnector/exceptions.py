"""Exceptions."""


class AuthenticationException(BaseException):
    """Raise this when there is an authentication issue."""


class BadMessageException(BaseException):
    """Raise this when a bad message is sent."""


class BadRequestException(BaseException):
    """Raise this when a bad request is sent."""


class ConnectionClosedException(BaseException):
    """Raise this when connection is closed."""


class ConnectionErrorException(BaseException):
    """Raise this when error connecting."""
