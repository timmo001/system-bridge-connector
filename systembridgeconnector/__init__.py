"""System Bridge Connector."""

from ._version import __version__  # re-export for convenience

__all__ = ["__version__"]
