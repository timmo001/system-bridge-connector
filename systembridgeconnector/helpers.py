"""Helpers."""

from collections.abc import Callable
import datetime as dt
from datetime import UTC, datetime, timedelta
from functools import partial

from systembridgemodels.const import ModulesData
from systembridgemodels.modules.cpu import PerCPU
from systembridgemodels.modules.displays import Display
from systembridgemodels.modules.gpus import GPU

utcnow = partial(dt.datetime.now, UTC)


def battery_time_remaining(data: ModulesData) -> datetime | None:
    """Return the battery time remaining."""
    if data.battery and (battery_time := data.battery.time_remaining) is not None:
        return utcnow() + timedelta(seconds=battery_time)
    return None


def camera_in_use(data: ModulesData) -> bool | None:
    """Return if any camera is in use."""
    if data.system and data.system.camera_usage is not None:
        return len(data.system.camera_usage) > 0
    return None


def cpu_speed(data: ModulesData) -> float | None:
    """Return the CPU speed."""
    if (
        data.cpu
        and (cpu_frequency := data.cpu.frequency) is not None
        and (cpu_frequency.current) is not None
    ):
        return round(cpu_frequency.current / 1000, 2)
    return None


def with_per_cpu(func) -> Callable:
    """Wrap a function to ensure per CPU data is available."""

    def wrapper(data: ModulesData, index: int) -> float | None:
        """Wrap a function to ensure per CPU data is available."""
        if data.cpu and data.cpu.per_cpu is not None and index < len(data.cpu.per_cpu):
            return func(data.cpu.per_cpu[index])
        return None

    return wrapper


@with_per_cpu
def cpu_power_per_cpu(per_cpu: PerCPU) -> float | None:
    """Return CPU power per CPU."""
    return per_cpu.power


@with_per_cpu
def cpu_usage_per_cpu(per_cpu: PerCPU) -> float | None:
    """Return CPU usage per CPU."""
    return per_cpu.usage


def with_display(func) -> Callable:
    """Wrap a function to ensure a Display is available."""

    def wrapper(data: ModulesData, index: int) -> Display | None:
        """Wrap a function to ensure a Display is available."""
        if data.displays and index < len(data.displays):
            return func(data.displays[index])
        return None

    return wrapper


@with_display
def display_resolution_horizontal(display: Display) -> int | None:
    """Return the Display resolution horizontal."""
    return display.resolution_horizontal


@with_display
def display_resolution_vertical(display: Display) -> int | None:
    """Return the Display resolution vertical."""
    return display.resolution_vertical


@with_display
def display_refresh_rate(display: Display) -> float | None:
    """Return the Display refresh rate."""
    return display.refresh_rate


def with_gpu(func) -> Callable:
    """Wrap a function to ensure a GPU is available."""

    def wrapper(data: ModulesData, index: int) -> GPU | None:
        """Wrap a function to ensure a GPU is available."""
        if data.gpus and index < len(data.gpus):
            return func(data.gpus[index])
        return None

    return wrapper


@with_gpu
def gpu_core_clock_speed(gpu: GPU) -> float | None:
    """Return the GPU core clock speed."""
    return gpu.core_clock


@with_gpu
def gpu_fan_speed(gpu: GPU) -> float | None:
    """Return the GPU fan speed."""
    return gpu.fan_speed


@with_gpu
def gpu_memory_clock_speed(gpu: GPU) -> float | None:
    """Return the GPU memory clock speed."""
    return gpu.memory_clock


@with_gpu
def gpu_memory_free(gpu: GPU) -> float | None:
    """Return the free GPU memory."""
    return gpu.memory_free


@with_gpu
def gpu_memory_used(gpu: GPU) -> float | None:
    """Return the used GPU memory."""
    return gpu.memory_used


@with_gpu
def gpu_memory_used_percentage(gpu: GPU) -> float | None:
    """Return the used GPU memory percentage."""
    if (gpu.memory_used) is not None and (gpu.memory_total) is not None:
        return round(gpu.memory_used / gpu.memory_total * 100, 2)
    return None


@with_gpu
def gpu_power_usage(gpu: GPU) -> float | None:
    """Return the GPU power usage."""
    return gpu.power_usage


@with_gpu
def gpu_temperature(gpu: GPU) -> float | None:
    """Return the GPU temperature."""
    return gpu.temperature


@with_gpu
def gpu_usage_percentage(gpu: GPU) -> float | None:
    """Return the GPU usage percentage."""
    return gpu.core_load


def memory_free(data: ModulesData) -> float | None:
    """Return the free memory."""
    if (
        data.memory
        and (virtual := data.memory.virtual) is not None
        and (free := virtual.free) is not None
    ):
        return round(free / 1000**3, 2)
    return None


def memory_used(data: ModulesData) -> float | None:
    """Return the used memory."""
    if (
        data.memory
        and (virtual := data.memory.virtual) is not None
        and (used := virtual.used) is not None
    ):
        return round(used / 1000**3, 2)
    return None


def partition_usage(
    data: ModulesData,
    device_index: int,
    partition_index: int,
) -> float | None:
    """Return the used memory."""
    if (
        data.disks
        and (devices := data.disks.devices) is not None
        and device_index < len(devices)
        and (partitions := devices[device_index].partitions) is not None
        and partition_index < len(partitions)
        and (usage := partitions[partition_index].usage) is not None
    ):
        return usage.percent
    return None
