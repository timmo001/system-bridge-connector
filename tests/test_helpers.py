"""Test the helpers module."""

from datetime import UTC, datetime
from unittest.mock import patch

import pytest
from syrupy.assertion import SnapshotAssertion

from systembridgeconnector.helpers import (
    battery_time_remaining,
    camera_in_use,
    cpu_power_per_cpu,
    cpu_speed,
    cpu_usage_per_cpu,
    display_refresh_rate,
    display_resolution_horizontal,
    display_resolution_vertical,
    gpu_core_clock_speed,
    gpu_fan_speed,
    gpu_memory_clock_speed,
    gpu_memory_free,
    gpu_memory_used,
    gpu_memory_used_percentage,
    gpu_power_usage,
    gpu_temperature,
    gpu_usage_percentage,
    memory_free,
    memory_used,
    partition_usage,
)
from systembridgemodels.modules import ModulesData

EMPTY_MODULES_DATA = ModulesData()


@pytest.mark.asyncio
async def test_battery_time_remaining(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test battery time remaining."""
    with patch(
        "systembridgeconnector.helpers.utcnow",
        return_value=datetime(2024, 1, 1, 0, 0, 0, 0, tzinfo=UTC),
    ):
        assert battery_time_remaining(mock_modules_data) == snapshot
        assert battery_time_remaining(EMPTY_MODULES_DATA) is None


@pytest.mark.asyncio
async def test_camera_in_use(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test camera in use."""
    assert camera_in_use(mock_modules_data) == snapshot
    assert camera_in_use(EMPTY_MODULES_DATA) is None


@pytest.mark.asyncio
async def test_cpu_speed(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test CPU speed."""
    assert cpu_speed(mock_modules_data) == snapshot
    assert cpu_speed(EMPTY_MODULES_DATA) is None


@pytest.mark.asyncio
async def test_cpu_power_per_cpu(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test CPU power per CPU."""
    assert cpu_power_per_cpu(mock_modules_data, 0) == snapshot
    assert cpu_power_per_cpu(EMPTY_MODULES_DATA, 0) is None


@pytest.mark.asyncio
async def test_cpu_usage_per_cpu(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test CPU usage per CPU."""
    assert cpu_usage_per_cpu(mock_modules_data, 0) == snapshot
    assert cpu_usage_per_cpu(EMPTY_MODULES_DATA, 0) is None


@pytest.mark.asyncio
async def test_display_resolution_horizontal(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test display resolution horizontal."""
    assert display_resolution_horizontal(mock_modules_data, 0) == snapshot
    assert display_resolution_horizontal(EMPTY_MODULES_DATA, 0) is None


@pytest.mark.asyncio
async def test_display_resolution_vertical(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test display resolution vertical."""
    assert display_resolution_vertical(mock_modules_data, 0) == snapshot
    assert display_resolution_vertical(EMPTY_MODULES_DATA, 0) is None


@pytest.mark.asyncio
async def test_display_refresh_rate(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test display refresh rate."""
    assert display_refresh_rate(mock_modules_data, 0) == snapshot
    assert display_refresh_rate(EMPTY_MODULES_DATA, 0) is None


@pytest.mark.asyncio
async def test_gpu_core_clock_speed(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test GPU core clock speed."""
    assert gpu_core_clock_speed(mock_modules_data, 0) == snapshot
    assert gpu_core_clock_speed(EMPTY_MODULES_DATA, 0) is None


@pytest.mark.asyncio
async def test_gpu_fan_speed(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test GPU fan speed."""
    assert gpu_fan_speed(mock_modules_data, 0) == snapshot
    assert gpu_fan_speed(EMPTY_MODULES_DATA, 0) is None


@pytest.mark.asyncio
async def test_gpu_memory_clock_speed(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test GPU memory clock speed."""
    assert gpu_memory_clock_speed(mock_modules_data, 0) == snapshot
    assert gpu_memory_clock_speed(EMPTY_MODULES_DATA, 0) is None


@pytest.mark.asyncio
async def test_gpu_memory_free(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test GPU memory free."""
    assert gpu_memory_free(mock_modules_data, 0) == snapshot
    assert gpu_memory_free(EMPTY_MODULES_DATA, 0) is None


@pytest.mark.asyncio
async def test_gpu_memory_used(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test GPU memory used."""
    assert gpu_memory_used(mock_modules_data, 0) == snapshot
    assert gpu_memory_used(EMPTY_MODULES_DATA, 0) is None


@pytest.mark.asyncio
async def test_gpu_memory_used_percentage(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test GPU memory used percentage."""
    assert gpu_memory_used_percentage(mock_modules_data, 0) == snapshot
    assert mock_modules_data.gpus
    mock_modules_data.gpus[0].memory_used = None
    assert gpu_memory_used_percentage(mock_modules_data, 0) is None
    assert gpu_memory_used_percentage(EMPTY_MODULES_DATA, 0) is None


@pytest.mark.asyncio
async def test_gpu_power_usage(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test GPU power usage."""
    assert gpu_power_usage(mock_modules_data, 0) == snapshot
    assert gpu_power_usage(EMPTY_MODULES_DATA, 0) is None


@pytest.mark.asyncio
async def test_gpu_temperature(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test GPU temperature."""
    assert gpu_temperature(mock_modules_data, 0) == snapshot
    assert gpu_temperature(EMPTY_MODULES_DATA, 0) is None


@pytest.mark.asyncio
async def test_gpu_usage_percentage(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test GPU usage percentage."""
    assert gpu_usage_percentage(mock_modules_data, 0) == snapshot
    assert gpu_usage_percentage(EMPTY_MODULES_DATA, 0) is None


@pytest.mark.asyncio
async def test_memory_free(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test memory free."""
    assert memory_free(mock_modules_data) == snapshot
    assert memory_free(EMPTY_MODULES_DATA) is None


@pytest.mark.asyncio
async def test_memory_used(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test memory used."""
    assert memory_used(mock_modules_data) == snapshot
    assert memory_used(EMPTY_MODULES_DATA) is None


@pytest.mark.asyncio
async def test_partition_usage(
    snapshot: SnapshotAssertion,
    mock_modules_data: ModulesData,
) -> None:
    """Test partition usage."""
    assert partition_usage(mock_modules_data, 0, 0) == snapshot
    assert partition_usage(EMPTY_MODULES_DATA, 0, 0) is None
