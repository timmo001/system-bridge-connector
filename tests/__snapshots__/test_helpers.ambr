# serializer version: 1
# name: test_battery_time_remaining
  datetime.datetime(2024, 1, 1, 0, 0, 12, tzinfo=datetime.timezone.utc)
# ---
# name: test_camera_in_use
  True
# ---
# name: test_cpu_power_per_cpu
  50.0
# ---
# name: test_cpu_speed
  0.0
# ---
# name: test_cpu_usage_per_cpu
  20.0
# ---
# name: test_display_refresh_rate
  60.0
# ---
# name: test_display_resolution_horizontal
  1920
# ---
# name: test_display_resolution_vertical
  1080
# ---
# name: test_gpu_core_clock_speed
  1000.0
# ---
# name: test_gpu_fan_speed
  100.0
# ---
# name: test_gpu_memory_clock_speed
  1000.0
# ---
# name: test_gpu_memory_free
  1000.0
# ---
# name: test_gpu_memory_used
  1000.0
# ---
# name: test_gpu_memory_used_percentage
  100.0
# ---
# name: test_gpu_power_usage
  100.0
# ---
# name: test_gpu_temperature
  100.0
# ---
# name: test_gpu_usage_percentage
  100.0
# ---
# name: test_memory_free
  0.0
# ---
# name: test_memory_used
  0.0
# ---
# name: test_partition_usage
  40.2
# ---
